pkg/querysharding/util.go

```diff
@@ -0,0 +1,76 @@
+package querysharding
+
+import (
+	"encoding/base64"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+const (
+	CortexShardByLabel = "__CORTEX_SHARD_BY__"
+)
+
+var (
+	buffers = sync.Pool{New: func() interface{} {
+		b := make([]byte, 0, 100)
+		return &b
+	}}
+)
+
+func InjectShardingInfo(query string, shardInfo *storepb.ShardInfo) (string, error) {
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return "", err
+	}
+	b, err := shardInfo.Marshal()
+	if err != nil {
+		return "", err
+	}
+	parser.Inspect(expr, func(n parser.Node, _ []parser.Node) error {
+		if selector, ok := n.(*parser.VectorSelector); ok {
+			selector.LabelMatchers = append(selector.LabelMatchers, &labels.Matcher{
+				Type:  labels.MatchEqual,
+				Name:  CortexShardByLabel,
+				Value: base64.StdEncoding.EncodeToString(b),
+			})
+		}
```

Comment:
I would really try to avoid the additional label. 
https://github.com/prometheus/prometheus/blob/main/storage/interface.go#L174 
I think the query hints is probably the best place for this kind of information, though seems not supported right now.

Reply:
What i was actually thinking actually was to propose to prometheus some kind of extensible `queryops` as it extract things from the request.

https://github.com/prometheus/prometheus/blob/35d6813963b3f82c5d07f3602cc5e6decd12f53e/web/api/v1/api.go#L443-L447
https://github.com/prometheus/prometheus/blob/35d6813963b3f82c5d07f3602cc5e6decd12f53e/web/api/v1/api.go#L169-L170

Thanos was able to implement this without the extra label because it "forked" the api from prometheus and modified it but i don't think this is ideal as well.

Because of that I decided to go with the extra label. Any label that starts with `__` is [considered internal](https://prometheus.io/docs/concepts/data_model/) and so we can in theory use them. 

Can you elaborate more about why this would be a problem?

Reply:
> What i was actually thinking actually was to propose to prometheus some kind of extensible queryops as it extract things from the request (queryops become an interface and we could pass the extractFunction to the engine).

Right I think this is the right way to push forward as sharding info is part of the query metadata semanticly.

> Thanos was able to implement this without the extra label because it "forked" the api from prometheus and modified it but i don't think this is ideal as well.

I am not comparing it with Thanos. Store APIs have a lot other use cases as well.

Overall, I think it is fine to keep the current way using an additional label since we don't have other choices. Ideally using queryops would be the best

Reply:
@alanprot Can I ask why we don't just extend `client.QueryRequest` protobuf to have the query sharding info?

Reply:
The problem is that i can indeed inject anything on the QueryRequest, but i cannot extract those info on the other side.

We are using the prometheus api file to handle those requests and prometheus currently had no way to allow us to extract extra info from the request. See https://github.com/prometheus/prometheus/blob/c1b669bf9b0b1286ece53c977262089119783105/web/api/v1/api.go#L381-L432